engine/src/main/java/org/hibernate/validator/internal/engine/ConstraintViolationImpl.java

```diff
@@ -228,7 +228,7 @@ public Object getExecutableReturnValue() {
 
 	@Override
 	public <C> C getDynamicPayload(Class<C> type) {
-		if ( type.isAssignableFrom( this.dynamicPayload.getClass() ) ) {
+		if ( dynamicPayload != null && type.isAssignableFrom( this.dynamicPayload.getClass() ) ) {
```

Comment:
Looks very good, thanks!

Minor comment: could you remove the `this.` occurrences in front of `dynamicPayload` in the existing code of this method and amend your commit? We should be consistent and I see no reason to keep `this.`.

Additional bonus point for being in line with our checkstyle rules!